### PR TITLE
feat(rpc): new endpoint for block validation status

### LINKED ISSUE #1128: add rpc endpoint to check validation status

### PR DESCRIPTION
**Pull request by srene**
# PR Standards

## Opening a pull request should be able to meet the following requirements

--

PR naming convention: https://hackmd.io/@nZpxHZ0CT7O5ngTp0TP9mg/HJP_jrm7A

---

Close #1128

<-- Briefly describe the content of this pull request -->

For Author:

- [ ]  Targeted PR against correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to Github issue with discussion and accepted design
- [ ]  Targets only one github issue
- [ ]  Wrote unit and integration tests
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

---

For Reviewer:

- [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  confirmed all author checklist items have been addressed

---

After reviewer approval:

- [ ]  In case targets main branch, PR should be squashed and merged.
- [ ]  In case PR targets a release branch, PR should be rebased.

### CODE REVIEW

File: rpc/client/client.go [hunk 1]
```diff
@@ -817,6 +821,24 @@ func (c *Client) CheckTx(ctx context.Context, tx tmtypes.Tx) (*ctypes.ResultChec
 	return &ctypes.ResultCheckTx{ResponseCheckTx: *res}, nil
 }
 
+// BlockValidated returns the settlement validation status for a specific height: 0: Not validated, 1: Validated from P2P, 2: Validated from settlement, -1: node is still syncing.
```
Comment (omritoptix) [inline rpc/client/client.go:824]: should probably an enum vs hardcoded

File: rpc/client/client.go [hunk 2]
```diff
@@ -53,6 +61,10 @@ type Client struct {
 	genChunks []string
 }
 
+type ResultBlockValidated struct {
+	Result int
```
Comment (danwt) [inline rpc/client/client.go:65]: why not use the custom type here?

File: rpc/json/service.go [hunk 3]
```diff
@@ -193,6 +194,7 @@ func (s *service) NetInfo(req *http.Request, args *netInfoArgs) (*ctypes.ResultN
 }
 
 func (s *service) BlockchainInfo(req *http.Request, args *blockchainInfoArgs) (*ctypes.ResultBlockchainInfo, error) {
+	fmt.Println(args)
```
Comment (zale144) [inline rpc/json/service.go:197]: remove

Comment (srene) [reply rpc/json/service.go]: done

**Comment by srene** [comment]
> I'm not exactly sure in what context this is used, but might it be desirable for the caller to also pass the header hash to check that the block the caller refers to is the same one the node is referring to?
> 
> To stop the caller and the callee confusing two different blocks with the same height
> 
> Basically the caller wants to know he has the same block the verifier has? or not?

that is a good point but i dont think that the software that requires this rpc has the block itself with the hash, just the height.

File: rpc/client/client.go [hunk 4]
```diff
@@ -817,6 +829,24 @@ func (c *Client) CheckTx(ctx context.Context, tx tmtypes.Tx) (*ctypes.ResultChec
 	return &ctypes.ResultCheckTx{ResponseCheckTx: *res}, nil
 }
 
+// BlockValidated returns the settlement validation status for a specific height: 0: Not validated, 1: Validated from P2P, 2: Validated from settlement, -1: node is still syncing.
+func (c *Client) BlockValidated(ctx context.Context, height *int64) (*ResultBlockValidated, error) {
```
Comment (danwt) [inline rpc/client/client.go:833]: why is height a pointer?

**Review by danwt — CHANGES_REQUESTED** [review]

File: rpc/client/client.go (hunk 4 above)
Comment (srene) [reply rpc/client/client.go:833]: i think its the common way used in all endpoints. it can be useful  to check if height its set in the input params.

File: rpc/client/client.go [hunk 5]
```diff
@@ -817,6 +829,25 @@ func (c *Client) CheckTx(ctx context.Context, tx tmtypes.Tx) (*ctypes.ResultChec
...
+	// invalid height
+	if *height < 0 || height == nil {
```
Comment (zale144) [inline rpc/client/client.go:835]: this can panic. for short circuit evaluation order is important, the OR is evaluated from left to right

```suggestion
	if height == nil || *height < 0 {
```

File: rpc/client/client_test.go [hunk 6]
```diff
@@ -295,6 +295,83 @@ func TestGetBlock(t *testing.T) {
...
+
+	tests := []struct {
+		name            string
+		validatedHeight uint64
+		nodeHeight      uint64
+		submittedHeight uint64
+		queryHeight     int64
+		result          client.ValidationStatus
+	}{
+		{
```
Comment (zale144) [inline rpc/client/client_test.go:318]: would it make sense to add a test case with `queryHeight: nil` ?

File: settlement/settlement.go [hunk 7]
```diff
@@ -32,6 +34,15 @@ type BatchMetaData struct {
 	DA *da.DASubmitMetaData
 }
 
+type BlockDescriptor struct {
```
Comment (zale144) [inline settlement/settlement.go:37]: is this a duplicate?

**Review by zale144 — CHANGES_REQUESTED** [review]

File: rpc/client/client.go (hunk 5 above)
Comment (srene) [reply rpc/client/client.go]: fixed

File: rpc/client/client_test.go (hunk 6 above)
Comment (srene) [reply rpc/client/client_test.go:320]: test added

File: settlement/settlement.go (hunk 7 above)
Comment (srene) [reply settlement/settlement.go]: yes, removed

File: rpc/client/client.go [hunk 8]
```diff
@@ -36,6 +36,14 @@ const (
 	subscribeTimeout = 5 * time.Second
 )
 
+type ValidationStatus int
+
+const (
+	BlockNotValidated = iota
+	BlockValidated
+	BlockValidatedWithSL
+)
```
Comment (omritoptix) [inline rpc/client/client.go:45]: imo better to stick to p2p as `BlockValidated` could be amibigous. 

```suggestion
type BlockValidationStatus int

const (
	NotValidated = iota
	P2PValidated
	SLValidated
)
```

**Review by omritoptix — APPROVED** [review]
